File: test/beta.c
#include <stddef.h>

// testing alternated error scheme
//
// error scheme: 
// positive status with 0 as default, negative numbers are errors

const char* str = NULL;

typedef enum {
    beta_excellent = 2,
    beta_good = 1,
    beta_ok = 0,
    beta_failure = -1,
    beta_invalid_arg = -2,
    beta_unexpected = -3,
} beta_status;

beta_status beta_accept_string(const char* s) {
    str = s;
    return beta_excellent;
}

beta_status beta_get_char(size_t index, char* out0) {
    if (str) {
        *out0 = str[index];
        return beta_ok;
    } else {
        return beta_unexpected;
    }
}

typedef union {
    int integer;
    double percentage;
} beta_union;

beta_status beta_accept_union(const beta_union* arg0, beta_union* out0) {
    out0->integer = arg0->integer + 1;
    return beta_ok;
}

typedef struct {
    int x;
    int y;
} beta_point;

beta_status beta_get_mood(void) {
    return beta_good;
}

beta_status beta_reject_arg(int arg0) {
    return beta_invalid_arg;
}

File: test/beta.zig
const std = @import("std");
const api_translator = @import("api-translator.zig");
const c = @cImport({
    @cInclude("beta.c");
});

pub const Error = error{
    Failure,
    InvalidArg,
    Unexpected,
};
pub const Status = enum(c_int) {
    excellent = 2,
    good = 1,
    ok = 0,
    failure = -1,
    invalid_arg = -2,
    unexpected = -3,
    _,
};

pub const acceptString: fn (
    s: [*:0]const u8,
) Error!void = c_to_zig.translate("beta_accept_string", true, true, .{});

pub const getChar: fn (
    index: usize,
) Error!u8 = c_to_zig.translate("beta_get_char", true, true, .{});

pub const Union = extern union {
    integer: c_int,
    percentage: f64,
};

pub const acceptUnion: fn (
    arg0: *const Union,
) Error!Union = c_to_zig.translate("beta_accept_union", true, true, .{});

pub const Point = extern struct {
    x: c_int,
    y: c_int,
};

pub const getMood: fn () Error!void = c_to_zig.translate("beta_get_mood", true, true, .{});

pub const rejectArg: fn (
    arg0: c_int,
) Error!void = c_to_zig.translate("beta_reject_arg", true, true, .{});

const c_to_zig = api_translator.Translator(.{
    .c_import_ns = c,
    .substitutions = &.{
        .{ .old = [*c]const c.beta_union, .new = *const Union },
        .{ .old = [*c]const u8, .new = [*:0]const u8 },
        .{ .old = c.beta_union, .new = Union },
    },
    .error_scheme = api_translator.BasicErrorScheme(Status, Error, .{
        .default_success_status = .excellent,
        .default_failure_status = .failure,
        .default_error = Error.Unexpected,
    }),
});

test {
    inline for (comptime std.meta.declarations(@This())) |decl| {
        _ = @field(@This(), decl.name);
    }
}
